src/io/parser.hpp

```diff
@@ -15,6 +15,135 @@
 
 namespace LightGBM {
 
+// define a str2num map class
+
+extern std::vector<Str2Num> maps;
+
+const char* Atof_and_map(const char* p, double* out, int idx) {
```

Comment:
>Atof_and_map [](http://example.com/codeflow?start=12&length=12)

I guess you don't need to write a string to float function by yourself. How about read out the string, trim the spaces and tabs, then check if it is a nan or infinite, then you can use the API to do that. I don't see any necessary to implement a str2num here